runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectRunnerTest.java

```diff
@@ -495,8 +495,36 @@ public void unreachable(ProcessContext c) {
                   }
                 }));
 
-    thrown.expectCause(isA(CoderException.class));
-    thrown.expectMessage("cannot encode a null Long");
+    thrown.expectCause(
+        new ThrowableMessageMatcher<CoderException>(equalTo("cannot encode a null Long")));
+    thrown.expectMessage("Exception occurred while executing transform Generate step");
+    p.run();
+  }
+
+  @Test
+  public void testUserException() throws Exception {
+    Pipeline p = getPipeline();
+    PCollection<Long> pcollection = p.apply("Step 1", Create.of(1))
+        .apply("Step 2", ParDo.of(new DoFn<Integer, Long>() {
+          @ProcessElement
+          public void processElement(ProcessContext c) {
+            throw new IllegalStateException("Custom exception element " + c.element());
+          }
+        })).setCoder(VarLongCoder.of());
+    pcollection
+        .apply("Unreachable",
+            ParDo.of(
+                new DoFn<Long, Long>() {
+                  @ProcessElement
+                  public void unreachable(ProcessContext c) {
+                    fail("Pipeline should fail to encode a null Long in VarLongCoder");
+                  }
+                }));
+
+    thrown.expectCause(
+        new ThrowableMessageMatcher<IllegalArgumentException>(equalTo("Custom exception "
+            + "element 1")));
+    thrown.expectMessage("Exception occurred while executing transform Step 2");
```

Comment:
Can you add a test involving a composite transform also?

Reply:
Added `testUserExceptionCompositeTransform`

Reply:
Done